### PR TITLE
Issue 2988: EPerson Password Replacement/addition with ADD Patch Operation instead of REPLACE

### LINKED ISSUE #2988: Cannot reset password for an EPerson created in Admin UI
**Describe the bug**
If you create a new EPerson via the Admin UI (http://localhost:4000/admin/access-control/epeople), then it's impossible to set or reset a password for that user.

**To Reproduce**
1. Login as an Admin
2. Create a new EPerson (Visit "Access Control" -> "People" -> "Add EPerson" -> fill out data -> "Submit")
3. Notice that the form doesn't let you set an initial password
4. Now, logout
5. Attempt to reset the new User's password by clicking "Have you forgotten your password?" and entering the user email address.  (NOTE: If you are running DSpace locally, you may need to set `mail.server.disabled=true` so that you can see the text of the generated email in your `dspace.log` file.)
6. Click on or visit the generated forgot email page (http://localhost:4000//forgot/[randomuuid])
7. Attempt to set a password.  You'll receive a 400 error that says: `Attempting to replace a non-existent value (netID).`

The problem appears to be in this method: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/patch/operation/EPersonPasswordReplaceOperation.java#L99

That method *requires* that a password exists before it lets you reset your password.  (It also throws an incorrect exception that mentions "netID" which has nothing to do with the actual error.)

**Expected behavior**
There *must* be a way to set a user's password after creating them from the Admin UI.  

One of the following MUST be possible:
1. Either the Admin should be allowed to initialize a password when creating the user account
2. OR, a random password should be initialized for the user account (and shown to the admin or emailed to the user)
3. OR, it should be possible to reset a password _even if_ a password was not already set.

My slight preference would be for option 3, as it's in line with DSpace 6 (and it should require no changes to the Angular UI).  I'm not against the other options if there's a clear reason why we do not want a password to be reset when it's empty.

### PR DESCRIPTION
**Pull request by Raf-atmire**
## References
* Fixes https://github.com/DSpace/DSpace/issues/2988

## Description
This PR changes the EPerson Password Replace Patch operation to an Add operation. It also removed the constraint that an EPerson had to have a Password prior to replacing it.
It fixes the issue where an EPerson without a password was never able to be assigned one with this Patch operation
This Add patch will also replace the password for an EPerson if it already had one

## Instructions for Reviewers

List of changes in this PR:
* Changed the EPerson Password Replace from a REPLACE patch operation to an ADD patch operation
* Changed tests and docs to reflect this change

How to test this PR:
* Create an EPerson without password
* Perform the Add Password Patch operation
* Try to login with the newly made password, ensure it works as expected

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java [hunk 1]
```diff
@@ -1410,14 +1411,14 @@ public void patchCertificateNonAdminUser() throws Exception {
     }
 
     @Test
-    public void patchPasswordMissingValue() throws Exception {
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:1413]: the scenario in the original test is loss. What happen if we try to set an empty password?

**Review by abollini — CHANGES_REQUESTED** [review]
hi @Raf-atmire 
thanks for this PR, the changes seem reasonable to me. I'm a bit worry about replacing the replace operation with the add instead to just add the new one (sorry for the pun). But as I'm not able to figure out a clear scenario where we take benefit of the replace I'm ok with that.

Please integrate the Test Case to deal with an attempt to set a null password

Before to merge we also need a PR for the rest contract to be update as well see https://github.com/DSpace/Rest7Contract/blame/main/epersons.md#L151

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java (hunk 1 above)
Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:1413]: I agree with @abollini that we should make sure to add a test here to prove what happens if you send an `ADD` with an empty password (I'd expect it to throw an error like before, but would be good to prove).

(Sidenote: @Raf-atmire  I've already done a quick code review & this is looking great to me, other then this one missing test.  I'll be waiting to submit my review though until I test it alongside the frontend PR)

**Review by tdonohue — APPROVED** [review]
@Raf-atmire : Tested this with the front-end and I can verify it resolves the bug reported in #2988

The code here looks good to me too, however I agree with @abollini that we need to add back in a test to verify what happens when an *empty* password is sent.   

I also agree that we need to update/correct the REST Contract for `/epersons` with these changes. I went ahead and created a PR for that: https://github.com/DSpace/Rest7Contract/pull/150

Once that missing test is added, I'm in favor of merging this as-is.

**Comment by Raf-atmire** [comment]
@tdonohue @abollini A test has been added by @MarieVerdonck that shows the behaviour when an ADD patch request is sent with a null password. It throws a BadRequest and the original password remains unaltered

**Review by tdonohue — APPROVED** [review]
👍 Thanks @Raf-atmire and @MarieVerdonck ! This looks good to me now.

**Comment by tdonohue** [comment]
@abollini : could you recheck this when you get a chance?  It looks (to me) like your feedback was resolved (and I created a corresponding REST Contract PR at https://github.com/DSpace/Rest7Contract/pull/150).    In my opinion this (and the corresponding Angular PR) are both ready to merge.

**Review by abollini — APPROVED** [review]

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java [hunk 2]
```diff
@@ -1411,49 +1412,70 @@ public void patchCertificateNonAdminUser() throws Exception {
...
                 .content(patchBody)
                 .contentType(MediaType.APPLICATION_JSON_PATCH_JSON))
-                        .andExpect(status().isBadRequest());
+                        .andExpect(status().isOk());
 
-        // login with original password
+        // login with new password => succeeds
         token = getAuthToken(ePerson.getEmail(), newPassword);
         getClient(token).perform(get("/api/"))
                         .andExpect(status().isOk());
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:1478]: I have approved this PR but I want to note that this test is not exactly what I would expect. It would be better to test for the authn/status endpoint to verify that we are really loggedin. It is not documented what should happen trying to access a public endpoint with an invalid token. @tdonohue do you agree? should we open a ticket for that? again this PR can be merged as is as this is unrelated with the PR

Comment (tdonohue) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/EPersonRestRepositoryIT.java:1478]: @Raf-atmire : @abollini is correct (and I overlooked this).  I don't think this test is correct (and I see you've reused this same test in your changes to `patchPasswordMissingValue()`).

Instead of doing a GET against `/api/`, we should be doing a GET against `/api/authn/status` and checking that it returns "authenticated=true". So, what you should be doing here is replace these two lines (starting with `getClient(` with somethin like this: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthenticationRestControllerIT.java#L211-L216

That'd be a more accurate test to verify that the login *works*.  I'm actually not sure if running a GET against `/api/` even requires authentication or not -- in fact based on the response of the demo REST API, it seems like that request would NOT require authentication: https://dspace7.4science.cloud/server/api/

So, if you could fix this check here and on line 1440 above, then this PR should be ready to merge.  Thanks!

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Raf-atmire : Not sure if you noticed [the comments above](https://github.com/DSpace/DSpace/pull/3041/files#r525560063).  A few tests are incorrect in that they are using the wrong endpoint (the endpoint they use will always succeed, even when unauthenticated).  Once you correct those tests, this will be ready to merge.

**Review by tdonohue — APPROVED** [review]
👍 Looks great now! Thanks @Raf-atmire and @MarieVerdonck !

I've also tested this backend with the frontend (https://github.com/DSpace/dspace-angular/pull/929)...both work well & the bug is resolved.  Thanks again!
